### DISCUSSION TITLE
Writing tests for functions returning mogwai `ViewBuilder`

### DISCUSSION THREAD

**Discussion by bryanjswift** [schell/mogwai#78] [Q&A] [answered]
I've got a "simple" function that results in a `ViewBuilder`. I'm trying to write some tests to verify the behavior of this function and the resulting `ViewBuilder` but I'm not sure if I'm approaching this the right way. It would be extremely useful for `ViewBuilder` to implement `PartialEq` but I recognize that might not be possible (or only possible when the `IsDomNode` also implements `PartialEq`).

The guts of this aren't too important but the function for which I'm writing tests is 

```rust
fn game_status(tx_game: &Transmitter<api::GameState>) -> ViewBuilder<HtmlElement> {
    let rx_game_status = Receiver::new();
    // Only send an update in to the `rx_game_status` if the status has changed
    tx_game.wire_filter_fold(
        &rx_game_status,
        api::GameStatus::OPEN,
        |status, game_state| {
            if *status == game_state.status {
                None
            } else {
                *status = game_state.status;
                Some(*status)
            }
        },
    );
    // Update the view whenever a new game status is received
    let rx_game_status_view = rx_game_status.branch_map(|status| match status {
        api::GameStatus::WON => Patch::Replace {
            index: 0,
            value: builder! { <h2>"You did the thing! 🥳"</h2> },
        },
        api::GameStatus::LOST => Patch::Replace {
            index: 0,
            value: builder! { <h2>"BOOM 💥"</h2> },
        },
        api::GameStatus::OPEN => Patch::Replace {
            index: 0,
            value: builder! { <span></span> },
        },
    });
    builder! {
        <slot name="game-status" patch:children=rx_game_status_view>
            <span></span>
        </slot>
    }
}
```

And I started writing tests like this

```rust
#[test]
fn stays_empty_on_open() {
    let tx = Transmitter::new();
    let view = game_status(&tx);
    tx.send(&api::GameState {
        id: uuid::Uuid::new_v4(),
        board: Vec::new(),
        status: api::GameStatus::OPEN,
    });
    assert_eq!(view.children.len(), 1);
    if let Some(ViewBuilder { element, .. }) = view.children.first() {
        assert_eq!(element.clone(), Some(String::from("span")));
    } else {
        assert!(false, "child was not <span>");
    }
    assert_eq!(view.patches.len(), 1);
}
```

This works ok but the next step would I think be validating the patches that come through and I think that's going to start to get quite difficult if I keep having to pick apart a `ViewBuilder`. I thought about writing some helper functions to do comparisons of the structs but I was curious if there was any better insight available for how to test?

**Comment by bryanjswift** [comment]
It occurred to me after writing this question that I could use the SSR capabilities to generate strings and then compare the resulting strings rather than pulling apart structs, which seems to solve my immediate troubles with initial rendering but still leaves the question of what to do about rendering the results of patches.

```rust
let ssr = View::from(view_builder).html_string();
assert_eq!(
    ssr,
    String::from("<slot name=\"game-status\"><span></span></slot>")
);
```

**Comment by bryanjswift** [comment, marked as answer]
Using the SSR capabilities via `View::html_string` does the trick (at least for now and this simple case) as long as the `View` is created before transmitting messages.

```rust
#[test]
fn booms_on_lost() {
    let tx = Transmitter::new();
    let builder = game_status(&tx);
    let ssr = View::from(builder);
    tx.send(&api::GameState {
        id: uuid::Uuid::new_v4(),
        board: Vec::new(),
        status: api::GameStatus::LOST,
    });
    assert_eq!(
        ssr.html_string(),
        String::from("<slot name=\"game-status\"><h2>BOOM 💥</h2></slot>")
    );
}
```

**Reply by schell** [reply]
Yes, the builder is not meant to hold the updated state of those transmitted messages. In other words the builder is not meant to be long-lived. You have the right idea in this iteration above.
